### PR TITLE
layers: Avoid redundant image layout checks

### PR DESCRIPTION
**Pull request by karl-lunarg**
When there are a large number of image layouts, keep track of the ones
that have been validated and avoid rechecking the same layouts.

This provides some performance improvement for applications providing
repeated layouts without significant impact to others.

### CODE REVIEW

File: layers/core_validation.cpp [hunk 1]
```diff
@@ -3135,10 +3135,15 @@ struct CommandBufferSubmitState {
                     // dynamic data isn't allowed in UPDATE_AFTER_BIND, so dynamicOffsets is always empty.
                     // This submit time not record time...
                     const bool record_time_validate = false;
-                    skip |= core->ValidateDescriptorSetBindingData(cb_node, set_node, dynamic_offsets, binding_info,
-                                                                   cmd_info.framebuffer, cmd_info.attachments.get(),
-                                                                   *cmd_info.subpasses.get(), record_time_validate,
-                                                                   function.c_str(), core->GetDrawDispatchVuid(cmd_info.cmd_type));
+                    layer_data::unordered_map<VkImageView, VkImageLayout> checked_layouts_map;
+                    layer_data::unordered_map<VkImageView, VkImageLayout> *checked_layouts = nullptr;
```
Comment (ncesario-lunarg) [inline layers/core_validation.cpp:3139]: Not sure, but it might make sense to collapse these into an "optional" (`zulauf::Optional`/`std::optional`). I'm guessing the unordered_map's constructor is essentially a no-op, but I _think_ using an optional would avoid any constructor work (aside from Optional's constructor) from being done in the `set_node->GetTotalDescriptorCount() <= cvdescriptorset::PrefilterBindRequestMap::kManyDescriptors_` case.

**Review by ncesario-lunarg — APPROVED** [review]
One comment/question, but LGTM!

File: layers/core_validation.cpp (hunk 1 above)
Comment (jeremyg-lunarg) [reply layers/core_validation.cpp:3139]: I don't think it'll matter performance, but this is a good idea from a clarity perspective.

**Review by jeremyg-lunarg — APPROVED** [review]

File: layers/core_validation.cpp (hunk 1 above)
Comment (karl-lunarg) [reply layers/core_validation.cpp:3139]: I did feel that this was a little clumsy and agree that `std::optional` would be more expressive.  But [I'm not sure that you can use reference parameters](https://stackoverflow.com/questions/64443556/modern-c-approach-for-providing-optional-arguments) in this situation.  (I assume you mean wrapping the map in `std::optional` and passing that type as a parameter to the verify function.)  The responses in that same article even say that doing this with a raw pointer is fine.  But I'll take a look at our `Optional` implementation to see what I can do.

Comment (jeremyg-lunarg) [reply layers/core_validation.cpp:3139]: In the past I've used boost::optional and done something similar to this; https://stackoverflow.com/questions/28955296/how-to-use-boostoptional-to-pass-optional-input-parmemters

Comment (ncesario-lunarg) [reply layers/core_validation.cpp:3139]: Yes, I was thinking of something like `Optional<layer_data::unordered_map<VkImageView, VkImageLayout>>` and then making the `checked_layouts` parameter a reference to `Optional<layer_data::unordered_map<VkImageView, VkImageLayout>>`. As Jeremy says, it probably makes no difference in this case.

tl;dr: I'm guessing the `layer_data::unordered_map`'s constructor doesn't do anything heavy. The idea with the optional is that _if_ the unordered_map's constructor does something potentially expensive (e.g., makes an allocation), that could be avoided.

e.g. (using constexpr conditionals for illustration),
- Pointer implementation where the object is not used: https://godbolt.org/z/K1Tavqccr (note the constructor is still called despite not being used)
- Pointer implementation where the object is used: https://godbolt.org/z/5sfGhvYzx
- Optional implementation where the object is not used: https://godbolt.org/z/r55a1sW5c (note the constructor is not called)
- Optional implementation where the object is used: https://godbolt.org/z/rs9ajcjMz

Comment (karl-lunarg) [reply layers/core_validation.cpp]: Yes, much nicer with `Optional` and an easy change.  Sandbox prototype: https://godbolt.org/z/jKx5GaW5E

Creating an empty `std::unordered_map` does involve an allocate and clear and I suspect it is the same when using the RH hashes.  But now this doesn't happen until I do the `emplace`, which doesn't happen unless the map will be used.  The perf tests don't show a lot of difference, though.  But still a good improvement for clarity.
